### PR TITLE
feat(log): Redacted user key-values printed in logs

### PR DESCRIPTION
**Pull request by Samunroyu**
https://github.com/apache/incubator-pegasus/issues/1575

User key-values will be redacted if encryption enabled.

### CODE REVIEW

File: src/base/pegasus_utils.h [hunk 1]
```diff
@@ -106,6 +107,18 @@ std::string c_escape_string(const T &src, bool always_escape = false)
 // ----------------------------------------------------------------------
 int c_unescape_string(const std::string &src, std::string &dest);
 
+template <class T>
+T redact_sensitive_string(const T &src, bool always_redacted = false)
```
Comment (acelyc111) [inline src/base/pegasus_utils.h:111]: It seems the second parameter is never used, how about removing it?

File: src/server/hotkey_collector.cpp [hunk 2]
```diff
@@ -196,7 +196,8 @@ inline void hotkey_collector::change_state_by_result()
         if (!_result.hot_hash_key.empty()) {
             change_state_to_finished();
             LOG_ERROR_PREFIX("Find the hotkey: {}",
-                             pegasus::utils::c_escape_string(_result.hot_hash_key));
+                             pegasus::utils::c_escape_string(
```
Comment (acelyc111) [inline src/server/hotkey_collector.cpp:199]: Is the c_escape_string always called along with redact_sensitive_string? How about unify to a single function?

File: src/base/pegasus_utils.cpp [hunk 3]
```diff
@@ -192,5 +195,41 @@ int c_unescape_string(const std::string &src, std::string &dest)
...
+
+template <>
+dsn::blob redact_sensitive_string(const dsn::blob &src)
+{
+    static dsn::blob rbb = dsn::blob::create_from_bytes(std::move("<redacted>"));
+    if (FLAGS_encrypt_data_at_rest) {
+        return rbb;
+    }
+
+    else {
```
Comment (acelyc111) [inline src/base/pegasus_utils.cpp:216]:
```suggestion
    } else {
```

File: src/base/pegasus_utils.cpp [hunk 4]
```diff
@@ -192,5 +195,41 @@ int c_unescape_string(const std::string &src, std::string &dest)
...
+
+template <>
+rocksdb::Slice redact_sensitive_string(const rocksdb::Slice &src)
+{
+    static rocksdb::Slice slice = rocksdb::Slice(std::move("<redacted>"));
+    if (FLAGS_encrypt_data_at_rest) {
+        return slice;
+    }
+
+    else {
```
Comment (acelyc111) [inline src/base/pegasus_utils.cpp:229]:
```suggestion
    } else {
```

File: src/base/pegasus_utils.h [hunk 5]
```diff
@@ -106,6 +118,9 @@ std::string c_escape_string(const T &src, bool always_escape = false)
 // ----------------------------------------------------------------------
 int c_unescape_string(const std::string &src, std::string &dest);
 
+template <typename T>
+T redact_sensitive_string(const T &src);
```
Comment (acelyc111) [inline src/base/pegasus_utils.h:122]: Is it possible to return a const reference type?

File: src/base/pegasus_utils.h [hunk 6]
```diff
@@ -106,6 +118,16 @@ std::string c_escape_string(const T &src, bool always_escape = false)
 // ----------------------------------------------------------------------
 int c_unescape_string(const std::string &src, std::string &dest);
 
+template <class T>
+std::string redact_sensitive_string(const T &src)
+{
+    if (FLAGS_encrypt_data_at_rest) {
+        return "<redacted>";
```
Comment (acelyc111) [inline src/base/pegasus_utils.h:125]: How about defining `"<redacted>"` as static constant varible?

**Review by acelyc111 — DISMISSED** [review]
@Samunroyu Thanks for the contribution!

**Comment by acelyc111** [comment]
@Samunroyu Could you please fix the IWYU issue, and add some unit tests for the 2 newly added functions?

**Review by acelyc111 — DISMISSED** [review]

**Review by acelyc111 — APPROVED** [review]

File: src/base/pegasus_utils.h [hunk 7]
```diff
@@ -29,6 +29,9 @@
 #include <vector>
 
 #include "utils/string_view.h"
+#include "utils/flags.h"
```
Comment (GehaFearless) [inline src/base/pegasus_utils.h:32]: Header files sorted alphabetically

**Review by acelyc111 — APPROVED** [review]

**Pull request merged**